neo4j/internal/bolt/bolt5.go

```diff
@@ -427,29 +386,31 @@ func (b *bolt5) discardStream(ctx context.Context) {
 		return
 	}
 
+	stream.discarding = true // pull response handler will discard any accumulated record for this stream
 	discarded := false
 	for {
-		_, batch, sum := b.receiveNext(ctx)
-		if batch {
-			if discarded {
-				// Response to discard, see below
-				b.streams.remove(stream)
-				b.checkStreams()
-				return
-			}
-			// Discard all! After this the next receive will get another batch
-			// as a response to the discard, we need to keep track of that we
-			// already sent a discard.
-			discarded = true
-			stream.fetchSize = -1
-			if b.state == bolt5StreamingTx && stream.qid != b.lastQid {
-				b.out.appendDiscardNQid(stream.fetchSize, stream.qid)
-			} else {
-				b.out.appendDiscardN(stream.fetchSize)
-			}
-			b.out.send(ctx, b.conn)
-		} else if sum != nil || b.err != nil {
-			// Stream is detached in receiveNext
+		if err := b.queue.receiveAll(ctx); err != nil {
+			return
+		}
+		if b.err != nil {
+			return
+		}
+		if stream.sum != nil || stream.err != nil {
+			return
+		}
+		if stream.endOfBatch && discarded {
```

Comment:
I think this check is not sufficient. Try writing a test to verify. This is the scenario I'm thinking of:
 1. Make a record stream with 4 records
 2. set fetch size to 2
 3. call `bolt.Next` 3 times (add another test with calling it 4 times)
 4. call `bolt.Discard`.
 
 The problem I suspect: there is nothing left to be discarded; everything has been `PULL`ed already, but the driver will enqueue a `DISCARD` anyway.

Reply:
if everything is pulled, then `stream.sum` should not be nil and will return earlier (see condition just above)?

Reply:
proven in 6ad32ecbd7d9d75434a1fc70e2cef1efbdff5cf6

Reply:
I yes! I failed to realize the interplay with the `receiveAll` and `if stream.sum != nil || stream.err != nil` above. :ok_hand: